console/console.go

```diff
@@ -0,0 +1,370 @@
+		return err
+	}
+	return nil
+}
+
+// Interactive starts an interactive user session, where input is propted from
+// the configured user prompter.
+func (c *Console) Interactive() {
+	var (
+		prompt    = c.prompt          // Current prompt line (used for multi-line inputs)
+		indents   = 0                 // Current number of input indents (used for multi-line inputs)
+		input     = ""                // Current user input
+		scheduler = make(chan string) // Channel to send the next prompt on and receive the input
+	)
+	// Start a goroutine to listen for promt requests and send back inputs
+	go func() {
+		for {
+			// Read the next user input
+			line, err := c.prompter.PromptInput(<-scheduler)
+			if err != nil {
+				// In case of an error, either clear the prompt or fail
+				if err == liner.ErrPromptAborted { // ctrl-C
+					prompt, indents, input = c.prompt, 0, ""
+					scheduler <- ""
+					continue
+				}
+				close(scheduler)
+				return
+			}
+			// User input retrieved, send for interpretation and loop
+			scheduler <- line
+		}
+	}()
+	// Monitor Ctrl-C too in case the input is empty and we need to bail
+	abort := make(chan os.Signal, 1)
+	signal.Notify(abort, os.Interrupt)
+
+	// Start sending prompts to the user and reading back inputs
+	for {
+		// Send the next prompt, triggering an input read and process the result
+		scheduler <- prompt
+		select {
+		case <-abort:
+			// User forcefully quite the console
+			fmt.Fprintln(c.printer, "caught interrupt, exiting")
+			return
+
+		case line, ok := <-scheduler:
+			// User input was returned by the prompter, handle special cases
+			if !ok || (indents <= 0 && exit.MatchString(line)) {
+				return
+			}
+			if onlyWhitespace.MatchString(line) {
+				continue
+			}
+			// Append the line to the input and check for multi-line interpretation
+			input += line + "\n"
+
+			indents = strings.Count(input, "{") + strings.Count(input, "(") - strings.Count(input, "}") - strings.Count(input, ")")
+			if indents <= 0 {
+				prompt = c.prompt
+			} else {
+				prompt = strings.Repeat("..", indents*2) + " "
+			}
+			// If all the needed lines are present, save the command and run
+			if indents <= 0 {
+				if len(input) > 0 && input[0] != ' ' && !passwordRegexp.MatchString(input) {
+					if command := strings.TrimSpace(input); len(c.history) == 0 || command != c.history[len(c.history)-1] {
+						c.history = append(c.history, command)
+						if c.prompter != nil {
+							c.prompter.AppendScrollHistory(command)
+						}
+					}
+				}
+				c.Evaluate(input)
+				input = ""
+			}
+		}
+	}
+}
+
+// Execute runs the JavaScript file specified as the argument.
+func (c *Console) Execute(path string) error {
+	return c.jsre.Exec(path)
+}
+
+// Stop cleans up the console and terminates the runtime envorinment.
+func (c *Console) Stop(graceful bool) error {
+	if err := ioutil.WriteFile(c.histPath, []byte(strings.Join(c.history, "\n")), 0600); err != nil {
```

Comment:
Fair point. I thought it would be enough to have it for new files, but you're right that we should set the correct permissions for old ones as well. Fixed and amended.